crates/astria-sequencer/src/accounts/ics20_withdrawal.rs

```diff
@@ -0,0 +1,140 @@
+};
+use penumbra_ibc::component::packet::{
+    IBCPacket,
+    SendPacketRead as _,
+    SendPacketWrite as _,
+    Unchecked,
+};
+use proto::native::sequencer::v1alpha1::{
+    asset,
+    asset::IbcAsset,
+    Address,
+    Ics20Withdrawal,
+};
+use tracing::instrument;
+
+use crate::{
+    accounts::state_ext::{
+        StateReadExt,
+        StateWriteExt,
+    },
+    transaction::action_handler::ActionHandler,
+};
+
+fn withdrawal_to_unchecked_ibc_packet(withdrawal: &Ics20Withdrawal) -> IBCPacket<Unchecked> {
+    let packet_data = withdrawal.to_fungible_token_packet_data();
+    let serialized_packet_data =
+        serde_json::to_vec(&packet_data).expect("can serialize FungibleTokenPacketData as JSON");
+
+    IBCPacket::new(
+        PortId::transfer(),
+        withdrawal.source_channel().clone(),
+        *withdrawal.timeout_height(),
+        withdrawal.timeout_time(),
+        serialized_packet_data,
+    )
+}
+
+#[async_trait::async_trait]
+impl ActionHandler for Ics20Withdrawal {
+    #[instrument(skip(self))]
+    async fn check_stateless(&self) -> Result<()> {
+        ensure!(self.timeout_time() != 0, "timeout time must be non-zero",);
+
+        // NOTE (from penumbra): we could validate the destination chain address as bech32 to
+        // prevent mistyped addresses, but this would preclude sending to chains that don't
+        // use bech32 addresses.
+        Ok(())
+    }
+
+    #[instrument(skip(self, state))]
+    async fn check_stateful<S: StateReadExt + 'static>(
+        &self,
+        state: &S,
+        from: Address,
+        _fee_asset_id: asset::Id,
+    ) -> Result<()> {
+        let packet: IBCPacket<Unchecked> = withdrawal_to_unchecked_ibc_packet(self);
+        state
+            .send_packet_check(packet)
+            .await
+            .context("packet failed send check")?;
+
+        let from_transfer_balance = state
+            .get_account_balance(from, self.denom().id())
+            .await
+            .context("failed getting `from` account balance for transfer")?;
+        ensure!(
+            from_transfer_balance >= self.amount(),
+            "insufficient funds for transfer"
+        );
+
+        Ok(())
+    }
+
+    #[instrument(skip(self, state))]
+    async fn execute<S: StateWriteExt>(
+        &self,
+        state: &mut S,
+        from: Address,
+        _fee_asset_id: asset::Id,
+    ) -> Result<()> {
+        let checked_packet = withdrawal_to_unchecked_ibc_packet(self).assume_checked();
+
+        let from_transfer_balance = state
+            .get_account_balance(from, self.denom().id())
+            .await
+            .context("failed getting `from` account balance for transfer")?;
+        ensure!(
+            from_transfer_balance >= self.amount(),
+            "insufficient funds for transfer"
+        );
+
+        state
+            .put_account_balance(
+                from,
+                self.denom().id(),
+                from_transfer_balance - self.amount(),
```

Comment:
You can do checked math && the check in one go:
```rust
let Some(new_balance) = from_transfer_balance.checked_sub(self.amount()) else {
    bail!("insufficient funds for transfer");
}
```
This addresses my ask for checked math while also defending against accidental refactor.